myhpom/models/state_requirement.py

```diff
@@ -57,10 +73,28 @@ def state_requirement_pre_save_receiver(sender, instance, **kwargs):
         raise ValidationError(errors)
 
     if instance.state is None:
-        if StateRequirement.objects.filter(state=None, position=instance.position).exists():
-            raise IntegrityError('(state, position) not unique: (None, %s)' % (instance.position,))
-        if StateRequirement.objects.filter(state=None, text=instance.text).exists():
-            raise IntegrityError('(state, text) not unique: (None, %s)' % (instance.text,))
+        # PostgreSQL doesn't raise IntegrityError on null state, so we have to.
+        if instance.id is None:
+            # creating a new instance
+            if StateRequirement.objects.filter(state=None, position=instance.position).exists():
+                raise IntegrityError(
+                    '(state, position) not unique: (None, %s)' % (instance.position,)
+                )
+            if StateRequirement.objects.filter(state=None, text=instance.text).exists():
+                raise IntegrityError('(state, text) not unique: (None, %s)' % (instance.text,))
+        else:
+            # saving an existing instance
+            qid = ~models.Q(id=instance.id)
+            if (
+                StateRequirement.objects.filter(qid)
+                .filter(state=None, position=instance.position)
+                .exists()
+            ):
+                raise IntegrityError(
+                    '(state, position) not unique: (None, %s)' % (instance.position,)
+                )
+            if StateRequirement.objects.filter(qid).filter(state=None, text=instance.text).exists():
+                raise IntegrityError('(state, text) not unique: (None, %s)' % (instance.text,))
```

Comment:
The 'position' appears to add some unnecessary complexity. As it is the admin interface is hard to use to change the ordering: one must assign some fictional value a StateRequirement, save, then assign the real values and save.

I suggest we remove the position field altogether and rely on pk order.

Neither position/no-position field makes changing the order easy (one requires juggling, the other deleting and re-adding), but I don't think we anticipate this being a regular activity. This model code and the admin interface would be much simpler to maintain and use w/o the position field.

Reply:
Now that we have admin, I agree with you. It's also possible that the admin user can/will change the order by copying/pasting to swap the text of existing entries, which is easier than the existing interface.